File: aoc15/day_01/day_01.ml
let input = Input.chars "input" in

let next floor n =
  let next_floor =
    match n with '(' -> floor + 1 | ')' -> floor - 1 | _ -> floor
  in
  (next_floor, next_floor)
in
let rec findindex pred lst =
  match lst with
  | [] -> -1
  | h :: t -> if pred h then 1 else 1 + findindex pred t
in
let first, floors = List.fold_left_map next 0 input in
let second = findindex (fun x -> x < 0) floors in

Printf.printf "First: %d\n" first;
Printf.printf "Second: %d\n" second

File: aoc15/go/day_01/day_01.go
package main

import (
	"fmt"
	"os"
)

func main() {
	input, _ := os.ReadFile("input")

	floor := 0
	second := 0
	for i, c := range input {
		switch c {
		case '(':
			floor++
		case ')':
			floor--
		}
		if floor < 0 && second == 0 {
			second = i + 1
		}
	}
	fmt.Println("First:", floor)
	fmt.Println("Second:", second)
}
